Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
+        @assert(!entry.asyncEvaluationWasEverTrue);
+
+        entry.asyncEvaluation = this.nextAsyncEvaluationValue++;
+        entry.asyncEvaluationWasEverTrue = true;
+        if (entry.pendingAsyncDependencies === 0)
+            this.executeAsyncModule(entry, fetcher);
     } else
-        return this.asyncModuleEvaluation(entry, fetcher, dependencies);
+        this.evaluate(entry.key, entry.moduleRecord, fetcher);
+
+    @assert(stack.@filter(e => e === entry).length === 1);
+    @assert(entry.dfsAncestorIndex <= entry.dfsIndex);
+
+    if (entry.dfsAncestorIndex === entry.dfsIndex) {
+        for (;;) {
+            var requiredModule = stack.@pop();
+            requiredModule.status = requiredModule.asyncEvaluation ? @ModuleStatusEvaluatingAsync : @ModuleStatusEvaluated;
+            requiredModule.cycleRoot = entry;
+            if (requiredModule === entry)
+                break;
+        }
+    }
+
+    return index;
 }
 
+// https://tc39.es/ecma262/#sec-execute-async-module
 @visibility=PrivateRecursive
-async function asyncModuleEvaluation(entry, fetcher, dependencies)
+async function executeAsyncModule(entry, fetcher)
 {
-    "use strict";
-
-    for (var i = 0, length = dependencies.length; i < length; ++i)
-        await this.moduleEvaluation(dependencies[i], fetcher);
+    @assert(entry.status === @ModuleStatusEvaluating || entry.status === @ModuleStatusEvaluatingAsync);
+    @assert(entry.hasTLA);
 
+    var { key, moduleRecord } = entry;
+    var sentValue = @undefined;
     var resumeMode = @GeneratorResumeModeNormal;
-    while (true) {
-        var awaitedValue = this.evaluate(entry.key, entry.module, fetcher, awaitedValue, resumeMode);
-        if (@getAbstractModuleRecordInternalField(entry.module, @abstractModuleRecordFieldState) == @GeneratorStateExecuting)
-            return;
 
-        try {
-            awaitedValue = await awaitedValue;
-            resumeMode = @GeneratorResumeModeNormal;
-        } catch (e) {
-            awaitedValue = e;
-            resumeMode = @GeneratorResumeModeThrow;
+    try {
+        for (;;) {
+            var result = this.evaluate(key, moduleRecord, fetcher, sentValue, resumeMode);
+            if (@getAbstractModuleRecordInternalField(moduleRecord, @abstractModuleRecordFieldState) === @GeneratorStateExecuting)
+                break;
+
+            try {
+                sentValue = await result;
+                resumeMode = @GeneratorResumeModeNormal;
+            } catch (error) {
+                sentValue = error;
+                resumeMode = @GeneratorResumeModeThrow;
+            }
+        }
+    } catch (error) {
+        @asyncModuleExecutionRejected(entry, error);
+        return;
+    }
+
+    this.asyncModuleExecutionFulfilled(entry, fetcher);
+}
+
+// https://tc39.es/ecma262/#sec-gather-available-ancestors
+@linkTimeConstant
+function gatherAvailableAncestors(entry, execList)
+{
+    for (var i = 0, { length } = entry.asyncParentModules; i < length; i++) {
```

Comment:
Do not use destructuring here. Overall, please do not use fancy features in builtin JS.
We need to be extra careful about what can be observable to users. And these fancy features make it super hard to handle.

Reply:
Removed destructuring and other fancy features from this file.